File: Data Structures & Algorithms/two-integer-sum-ii/submission-1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        
        if(numbers.size() == 2)
            return {1,2};

        int l = 0;
        int r = numbers.size() - 1;

        while(l < r) {
            int sum = numbers[l] + numbers[r];

            if(sum < target) {
                ++l;
            } else if(sum > target) {
                --r;
            } else {
                return {l+1, r+1};
            }

        }

        return {};

    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()

        def dfs(i, currlist, total):
            if total == target:
                res.append(currlist.copy())
                return
            
            for j in range(i, len(nums)):
                if total + nums[j] > target:
                    return

                currlist.append(nums[j])
                dfs(j, currlist, total + nums[j])
                currlist.pop()
                

        dfs(0, [], 0)
        return res